crates/dojo-core/src/database.cairo

```diff
@@ -17,79 +17,158 @@ mod utils_test;
 
 use index::WhereCondition;
 
-fn get(table: felt252, key: felt252, offset: u8, length: usize, layout: Span<u8>) -> Span<felt252> {
+// Could be replaced with a `KeyValues` with one value, 
+// but this allows us to avoid hashing, and is most common.
+#[derive(Copy, Drop, Serde)]
+struct MemberClause {
+    model: felt252,
+    member: felt252,
+    value: felt252,
+}
+
+#[derive(Copy, Drop, Serde)]
+struct CompositeClause {
+    operator: LogicalOperator,
+    clauses: Span<Clause>,
+}
+
+#[derive(Copy, Drop, Serde)]
+enum LogicalOperator {
+    And,
+}
+
+#[derive(Copy, Drop, Serde)]
+enum Clause {
+    Member: MemberClause,
+    Composite: CompositeClause,
+    All: felt252,
+}
+
+fn get(model: felt252, key: felt252, offset: u8, length: usize, layout: Span<u8>) -> Span<felt252> {
     let mut keys = ArrayTrait::new();
     keys.append('dojo_storage');
-    keys.append(table);
+    keys.append(model);
     keys.append(key);
     storage::get_many(0, keys.span(), offset, length, layout)
 }
 
-fn set(table: felt252, key: felt252, offset: u8, value: Span<felt252>, layout: Span<u8>) {
+fn set(model: felt252, key: felt252, offset: u8, value: Span<felt252>, layout: Span<u8>) {
     let mut keys = ArrayTrait::new();
     keys.append('dojo_storage');
-    keys.append(table);
+    keys.append(model);
     keys.append(key);
     storage::set_many(0, keys.span(), offset, value, layout);
 }
 
+/// Creates an entry in the database and adds it to appropriate indexes.
+/// # Arguments
+/// * `model` - The model to create the entry in.
+/// * `id` - id of the created entry.
+/// * `members` - The members to create an index on.
+/// * `offset` - The offset of the entry.
+/// * `value` - The value of the entry.
+/// * `layout` - The layout of the entry.
 fn set_with_index(
-    table: felt252, key: felt252, offset: u8, value: Span<felt252>, layout: Span<u8>
+    model: felt252,
+    key: felt252,
+    members: Span<felt252>,
+    offset: u8,
+    values: Span<felt252>,
+    layout: Span<u8>
 ) {
-    set(table, key, offset, value, layout);
-    index::create(0, table, key);
+    set(model, key, offset, values, layout);
+    index::create(0, model, key, 0); // create a record in index of all records
+
+    let mut idx = 0;
+    loop {
```

Comment:
Because key is a [hash of all member values](https://github.com/neotheprogramist/dojo/blob/696040ee05e11a19662aaba7454079d69c68d362/crates/dojo-core/src/world.cairo#L494-L496), when any of the them change, the object has a different key, and can't be accessed nor updated.

If it is a value that changes none of the indexes should be updated and none will be updated because of [this check](https://github.com/neotheprogramist/dojo/blob/696040ee05e11a19662aaba7454079d69c68d362/crates/dojo-core/src/database/index.cairo#L16-L18). That being said, will change it to check only once, for performance reasons.